### PR TITLE
Update /etc/os-release based on stack.toml content

### LINKED ISSUE #82: create-stack: update /etc/os-release
Currently, when using `jam create-stack` we have to manually update the values in `/etc/os-release` (e.g. [tiny stack](https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/e9a41436b06d3c3fa2960980ea98be36796ab3b6/stack/run/run.Dockerfile#L35) and we do something similar for the base and full stacks).

This file could be generated by `jam create-stack` rather than requiring users to put this in their `Dockerfile`.

`/etc/os-release` is present on all linux operating systems, so we could write it only when the `platform` contains `linux`.

### PR DESCRIPTION
**Pull request by modulo11**
## Summary

This change fixes #82 by reading the necessary adjustments for `/etc/os-release` from `stack.toml`. It is based on two mandatory fields `name` (pretty name of the stack) and `homepage` (already included). The additionally introduced fields `support-url` and `bug-report-url` are defaulted off from `homepage`. This replicates the capabilities currently implemented as part of the run-image [Dockerfile](https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/0ee5f28aaf767707840797b395c19c1a4d09809f/stack/run/run.Dockerfile#L35-L42) and a separate [os-release](https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/main/stack/run/files/os-release) file. Afterwards the stacks should be updated accordingly.

## Use Cases

- Dynamically update `/etc/os-release` with data coming from `stack.toml`.

## Checklist

* [X] I have viewed, signed, and submitted the Contributor License Agreement.
* [X] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [X] I have added an integration test, if necessary.
* [X] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

CC @phil9909

### CODE REVIEW

File: internal/ihop/definition.go
Comment (ryanmoran) [inline internal/ihop/definition.go:172]: Its maybe worth doing some kind of match on the homepage URL to check that its pointing to GitHub. If we move this code earlier (prior to the required fields check), then we can only these values if they are empty and matching `github.com`, and we can fail the parse call otherwise if they are truly required.

**Review by ryanmoran — COMMENTED** [review]
This is a great refactoring. I've just got one comment about how the new fields should be handled in the definition file. Its not clear to me if we should mark them as required, or just not update them in the layer creator if they are not set in the definition. I'm open to thoughts on this.

File: internal/ihop/definition.go
Comment (robdimsdale) [reply internal/ihop/definition.go:172]: I like the proposed behavior of having `name` and `homepage` as required and `support-url` and `bug-report-url` as optional in the `stack.toml` file, completing them based on `homepage` if they are omitted.

I also have a slight preference for checking that the `homepage` contains `github.com` before filling in the blanks. It seems easy to miss this when creating a stack whose `homepage` is not on github. I don't expect folks to read the code/docs before making a stack - I expect most people will copy an example `stack.toml` file from somewhere, and most of those existing stacks will have a `homepage` on github.

**Comment by robdimsdale** [comment]
As far as the `semver` label, I was thinking `minor`. However, I could see an argument for `major` given that we are requiring new fields in `stack.toml`. Thoughts @paketo-buildpacks/stacks-maintainers @ryanmoran ?

**Comment by modulo11** [comment]
> This is a great refactoring. I've just got one comment about how the new fields should be handled in the definition file. Its not clear to me if we should mark them as required, or just not update them in the layer creator if they are not set in the definition. I'm open to thoughts on this.

No strong opinion if they should both be required as well. We wanted to make the update on the tiny stack as minimal as possible. We'd also be fine to just copy only the data given from `stack.toml` (`name`, `homepage`, `support-url` or `bug-report-url`) into `/etc/os-release`.

After adding the check, if homepage contains `github.com`, it does not need to be mandatory any more.

File: internal/ihop/os_release_layer_creator.go
Comment (ryanmoran) [inline internal/ihop/os_release_layer_creator.go:93]: A couple thoughts on this section:
* Should we use `strconv.Quote`? https://pkg.go.dev/strconv#Quote
* If the values of the support and bug report URLs are empty, should we keep the original values?

Comment (robdimsdale) [reply internal/ihop/os_release_layer_creator.go:93]: I think it's fine to put empty strings for `support` and `bug_url`, as the only way you would get empty values is explicitly putting empty values in the `stack.toml`, right? Assume that's the only path, it seems reasonable to support the stack author's intention of leaving those fields blank.

File: internal/ihop/definition.go
Comment (robdimsdale) [inline internal/ihop/definition.go:166]: Did we want to fail if the `SupportURL`/`BugReportURL` are empty and the homepage doesn't contain `github.com`?

**Review by robdimsdale — COMMENTED** [review]
I have one outstanding question about whether we want to fail if the homepage doesn't contain `github.com` or silently pass.

Other than that - looks good.

**Comment by phil9909** [comment]
> As far as the `semver` label, I was thinking `minor`. However, I could see an argument for `major` given that we are requiring new fields in `stack.toml`. Thoughts @paketo-buildpacks/stacks-maintainers @ryanmoran ?

The following just came to my mind: We could make `name` and `homepage` optional and in the `internal/ihop/creator.go` only call `c.osReleaseLayerCreator.Create`, if the necessary values are present. That way it would be completely backwards compatible.

**Comment by modulo11** [comment]
> I have one outstanding question about whether we want to fail if the homepage doesn't contain `github.com` or silently pass.
> 
> Other than that - looks good.

We even thought about removing that `github.com` logic entirely and leave it up to the stack owners to provide this information. 
Once stacks are updated, e.g. remove the old `/etc/os-release` overwrite, fields like `homepage`, `bug-report-url` and `support-url` can be made mandatory.

WDYT?

File: internal/ihop/creator.go
Comment (ryanmoran) [inline internal/ihop/creator.go:122]:
```suggestion
	if def.containsOsReleaseOverwrites() {
```

File: internal/ihop/definition.go
Comment (ryanmoran) [inline internal/ihop/definition.go:203]:
```suggestion
func (d Definition) containsOsReleaseOverwrites() bool {
```

File: internal/ihop/os_release_layer_creator.go
Comment (ryanmoran) [inline internal/ihop/os_release_layer_creator.go:47]:
```suggestion
	buffer, err := overwriteOsRelease(content, createOsReleaseOverwrites(o.Def))
```

File: internal/ihop/os_release_layer_creator.go
Comment (ryanmoran) [inline internal/ihop/os_release_layer_creator.go:75]:
```suggestion
func createOsReleaseOverwrites(def Definition) map[string]string {
```

**Review by ryanmoran — COMMENTED** [review]
Short of a few typos, I am happy with this current implementation. @robdimsdale any other thoughts before we get this merged?

**Review by robdimsdale — DISMISSED** [review]

**Comment by robdimsdale** [comment]
I'm happy with this, and with the backwards-compatible changes we can label this: `minor`.

Thanks again for the PR @modulo11 @phil9909 @pbusko !

**Review by ryanmoran — APPROVED** [review]

**Review by robdimsdale — APPROVED** [review]

**Comment by robdimsdale** [comment]
Merged - thank you everyone for your contributions and for your ongoing effort keeping this branch up to date with `main` while we worked through the feedback.

**Pull request merged**
